tests/NavigationTests.elm

```diff
@@ -0,0 +1,156 @@
+        , update = programUpdate
+        , subscriptions = \_ -> Sub.none
+        , view = always <| div [] []
+        }
+        |> TestContext.start
+
+
+sampleProgramWithFlags : TestContext Model Msg
+sampleProgramWithFlags =
+    Navigation.programWithFlags UrlChange
+        { init = initWithStringFlags
+        , update = programUpdate
+        , subscriptions = \_ -> Sub.none
+        , view = always <| div [] []
+        }
+        |> TestContext.startWithFlags "foo"
+
+
+all : Test
+all =
+    describe "Navigation TestContext"
+        [ test "starting a navigation program" <|
+            \() ->
+                sampleProgram
+                    |> expectHref "https://elm.testable/"
+        , test "pushUrl" <|
+            \() ->
+                sampleProgram
+                    |> update (PushUrl "/foo")
+                    |> expectHref "https://elm.testable/foo"
+        , test "modifyUrl" <|
+            \() ->
+                sampleProgram
+                    |> update (ModifyUrl "/foo")
+                    |> expectHref "https://elm.testable/foo"
+        , test "back" <|
+            \() ->
+                sampleProgram
+                    |> update (PushUrl "/foo")
+                    |> update (PushUrl "/bar")
+                    |> update (Back 1)
+                    |> expectHref "https://elm.testable/foo"
+        , test "forward" <|
+            \() ->
+                sampleProgram
+                    |> update (PushUrl "/bar")
+                    |> update (PushUrl "/baz")
+                    |> update (Back 2)
+                    |> update (Forward 1)
+                    |> expectHref "https://elm.testable/bar"
+        , test "load" <|
+            \() ->
+                sampleProgram
+                    |> update (Load "/foo")
+                    |> expectHref "https://elm.testable/foo"
+        , test "refresh does nothing" <|
+            \() ->
+                sampleProgram
+                    |> update Reload
+                    |> done
+        , describe "navigation simulation"
+            [ test "simulates navigation for testing" <|
+                \() ->
+                    sampleProgram
+                        |> navigate "/qux"
+                        |> expectHref "https://elm.testable/qux"
+            , test "allows to go back" <|
+                \() ->
+                    sampleProgram
+                        |> navigate "/foo"
+                        |> navigate "/bar"
+                        |> update (Back 1)
+                        |> expectHref "https://elm.testable/foo"
+            , test "erases forward history" <|
+                \() ->
+                    sampleProgram
+                        |> navigate "/foo"
+                        |> update (Back 1)
+                        |> navigate "/baz"
+                        |> update (Back 100)
+                        |> update (Forward 1)
+                        |> expectHref "https://elm.testable/baz"
```

Comment:
It stays at the same place, that was what I was actually trying to test here, going back a hundred times to see how it behaves at the beginning of history, but perhaps would be better if we added a back/forward function like you said above to exercise the whole flow instead of calling the Msgs directly